hotupgrade-demo/sidecar/main.go

```diff
@@ -0,0 +1,276 @@
+/*
+Copyright 2021 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"k8s.io/klog"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"syscall"
+	"time"
+)
+
+var (
+	Version = ""
+)
+
+const (
+	unixSocket = "/unix-socket/migrate.sock"
+
+	// sidecar container version in container env(SIDECARSET_VERSION)
+	SidecarSetVersionEnvKey = "SIDECARSET_VERSION"
+	// container version env in the other sidecar container of the same hotupgrade sidecar(SIDECARSET_VERSION_ALT)
+	SidecarSetVersionAltEnvKey = "SIDECARSET_VERSION_ALT"
+)
+
+func main(){
+	isHotUpgrading, newer := isHotUpgradeProcess()
+	var tcpLn *net.TCPListener
+	var err error
+	// in hot upgrade process
+	if isHotUpgrading {
+		klog.Infof("start hot upgrade")
+		// Abnormal exit scenario of old version sidecar during hot upgrade
+		if !newer {
+			klog.Infof("older sidecar don't listen, and exit 0")
+			time.Sleep(time.Minute)
+			os.Exit(0)
+		}
+		tcpLn, err = triggerOldSidecarMigrate()
+		if err!=nil {
+			klog.Errorf("migrate older sidecar ListenFd failed, and will Listen new")
+		}
+	}
+	if tcpLn==nil {
+		tcpAddr, err := net.ResolveTCPAddr("tcp", ":9091")
+		if err!=nil {
+			klog.Errorf(err.Error())
+			return
+		}
+		tcpLn,err = net.ListenTCP("tcp", tcpAddr)
+		if err!=nil {
+			klog.Errorf(err.Error())
+			return
+		}
+	}
+
+	serveMux := http.NewServeMux()
+	serveMux.HandleFunc("/migrate", func(writer http.ResponseWriter, request *http.Request) {
+		err := migrateListenFd(tcpLn)
```

Comment:
can we demonstrate graceful shutdown of existing connections, or simply adding more comments here